### PR TITLE
Complete implementation of ObjectMetadata class.

### LINKED ISSUE #693: Complete the implementation of storage::ObjectMetadata.
Like for `storage::BucketMetadata` in #537, we need to complete the implementation.

### PR DESCRIPTION
**Pull request by coryan**
This fixes #693. The class can parse the full `storage#object`, as
described in:

https://cloud.google.com/storage/docs/json_api/v1/objects

I also introduced modifiers for all the "writable" fields. Some of
the refactoring touched BucketMetadata, this PR does *not*
complete the implementation of that class though.

### CODE REVIEW

File: google/cloud/storage/internal/metadata_parser_test.cc [hunk 1]
```diff
@@ -105,10 +102,71 @@ TEST(MetadataParserTest, ParseInvalidLongFieldType) {
...
+      "size": "42"
+})""";
+  auto json_object = nl::json::parse(text);
+  auto actual = ParseUnsignedLongField(json_object, "size");
+
+  EXPECT_EQ(42UL, actual);
+}
+
+/// @test Verify that we parse missing long values in JSON objects.
+TEST(MetadataParserTest, ParseMissinUnsignedLongField) {
```
Comment (houglum) [inline google/cloud/storage/internal/metadata_parser_test.cc]: s/Missin/Missing/

File: google/cloud/storage/object_metadata.cc [hunk 2]
```diff
@@ -20,38 +20,111 @@ namespace google {
...
-  result.generation_ =
-      internal::MetadataParser::ParseLongField(json, "generation");
+  result.cache_control_ = json.value("cacheControl", "");
+  result.component_count_ = internal::ParseLongField(json, "componentCount");
+  result.content_disposition_ = json.value("contentDisposition", "");
+  result.content_encoding_ = json.value("contentEncoding", "");
+  result.content_language_ = json.value("contentLanguage", "");
+  result.content_type_ = json.value("contentType", "");
+  result.crc32c_ = json.value("crc32c", "");
+  if (0 != json.count("customerEncryption")) {
```
Comment (houglum) [inline google/cloud/storage/object_metadata.cc]: Nit: Be consistent with which side of the `!=` operator you place the literal value (here, it's written as `0 != json.count(...)`, but above, it's written as `json.count(...) != 0`).

File: google/cloud/storage/object_metadata.h [hunk 3]
```diff
@@ -16,57 +16,161 @@
...
+  //@{
+  /// @name Accessors and modifiers to the `metadata` labels.
   std::size_t metadata_count() const { return metadata_.size(); }
   bool has_metadata(std::string const& key) const {
     return metadata_.end() != metadata_.find(key);
   }
   std::string const& metadata(std::string const& key) const {
     return metadata_.at(key);
   }
+  ObjectMetadata& delete_metadata(std::string const& key) {
```
Comment (houglum) [inline google/cloud/storage/object_metadata.h]: Probably worth adding a note that this succeeds (and does nothing) if the metadata item at the given key doesn't exist.

File: google/cloud/storage/internal/metadata_parser_test.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/metadata_parser_test.cc]: Done.

File: google/cloud/storage/object_metadata.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/object_metadata.cc]: Fixed.

File: google/cloud/storage/object_metadata.h (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/object_metadata.h]: Done.

**Pull request merged**